pretyping/evarconv.ml

```diff
@@ -1196,20 +1196,19 @@ let check_problems_are_solved env evd =
   | (pbty,env,t1,t2) as pb::_ -> error_cannot_unify env evd pb t1 t2
   | _ -> ()
 
+exception MaxUndefined of Evar.t * evar_info
+
 let max_undefined_with_candidates evd =
-  (* If evar were ordered with highest index first, fold_undefined
-     would be going decreasingly and we could use fold_undefined to
-     find the undefined evar of maximum index (alternatively,
-     max_bindings from ocaml 3.12 could be used); instead we traverse
-     the whole map *)
-  let l = Evd.fold_undefined
-    (fun evk ev_info evars ->
-        match ev_info.evar_candidates with
-        | None -> evars
-        | Some l -> (evk,ev_info,l)::evars) evd [] in
-  match l with
-  | [] -> None
-  | a::l -> Some (List.last (a::l))
+  (* Rely on the decreasing order to find the undefined evar of maximum index *)
+  let fold evk evi () = match evi.evar_candidates with
+  | None -> ()
+  | Some _ -> raise (MaxUndefined (evk, evi))
+  in
+  try
+    let () = Evar.Map.fold_right fold (Evd.undefined_map evd) () in
+    None
+  with MaxUndefined (evk, evi) ->
+    Some (evk, evi, Option.get evi.evar_candidates)
```

Comment:
I suggest you'd send the `evi.evar_candidates` in the Exception, as it has obtained by pattern matching in the `fold` function.